dashboard/app/assets/javascripts/client_state.js

```diff
@@ -180,4 +180,35 @@ function hasSeenVisualElement(visualElementType, visualElementId) {
   }
 }
 
+/**
+ * Queries the server for updated script progress for the current user and
+ * merges it into the current clientState.
+ *
+ * @param scriptId
+ * @returns jquery request object
+ */
+dashboard.clientState.refreshScriptProgressFromServer = function (scriptId) {
+  var req = $.post("/milestone/script_progress/" + scriptId);
+  req.done(function (responseJSON) {
```

Comment:
yea, JavaScript has a very well defined main threading model.  Basically, you can act like there's only one thread.  So there's no chance of the request finishing before the done gets added. However, if it did finish before hand the done would just get executed immediately.